### PR TITLE
Add backup and restore feature

### PR DESCRIPTION
**Pull request by Alain94W**
This branch contain the backup functionality, the restore is in progress

### CODE REVIEW

File: src/devices.js [hunk 1]
```diff
@@ -383,6 +386,80 @@ function assembleInstallSteps(steps) {
...
+              resolve();
+             }
+       }).catch (e=>{ 
+                     adb.shell("mount "+data_partition+" /data").then( (stdout)=>{
+                         utils.log.info("has been mounted");
+                         resolve();
+                     }).catch(e=>{reject("Unable to mount partition "+data_partition+" "+e)}); // mount
+                           //reject("Partition not mounted "+e);
+                    });//ls
+  });
```
Comment (NeoTheThird) [inline src/devices.js]: indentation

File: src/devices.js [hunk 2]
```diff
@@ -452,17 +532,12 @@ module.exports = {
                     resolve(option);
                   })
                   .catch(e =>
-                    reject(
-                      new Error("fetching system image channels failed: " + e)
-                    )
+                    reject("error fetching system image channels: " + e)
```
Comment (NeoTheThird) [inline src/devices.js]: Undo this.

File: src/devices.js [hunk 3]
```diff
@@ -452,17 +532,12 @@ module.exports = {
...
                   );
                 break;
               default:
                 reject(
-                  new Error(
-                    "unknown remote_values provider: " +
-                      option.remote_values.type
-                  )
+                  "unknown remote_values provider: " + option.remote_values.type
```
Comment (NeoTheThird) [inline src/devices.js]: Undo this.

File: src/html/modals/install.pug [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#install-modal.modal.fade(tabindex='-1', role='dialog')
```
Comment (NeoTheThird) [inline src/html/modals/install.pug]: delete this file again

File: src/html/modals/legacy-android.pug [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#legacy-android-modal.modal.fade(tabindex='-1', role='dialog')
```
Comment (NeoTheThird) [inline src/html/modals/legacy-android.pug]: delete this file again

File: src/html/index.pug [hunk 6]
```diff
@@ -32,20 +37,17 @@ html
           span#footer-bottom.text-muted Starting adb service
           span#footer-speed.text-muted
     // --- Modals ---
-    // low prio
+    include modals/install
+    include modals/connection-lost
+    include modals/oem-lock
+    include modals/legacy-android
+    include modals/low-power
```
Comment (NeoTheThird) [inline src/html/index.pug]: Undo this.

File: src/html/index.pug [hunk 7]
```diff
@@ -32,20 +37,17 @@ html
...
     include modals/select-device
     include modals/developer-mode
-    include modals/oem-lock
```
Comment (NeoTheThird) [inline src/html/index.pug]: Undo this.

File: src/html/index.pug [hunk 8]
```diff
@@ -32,20 +37,17 @@ html
...
     include modals/options
-    // medium prio
-    include modals/unlock
-    // high prio
+    include modals/new-update
```
Comment (NeoTheThird) [inline src/html/index.pug]: Undo this.

File: src/html/index.pug [hunk 9]
```diff
@@ -32,20 +37,17 @@ html
...
     include modals/udev
     include modals/windows-drivers
-    include modals/new-update
-    // errors
-    include modals/low-power
-    include modals/connection-lost
```
Comment (NeoTheThird) [inline src/html/index.pug]: Undo this.

File: src/html/modals/options.pug [hunk 10]
```diff
@@ -8,7 +8,6 @@
       .modal-body
         div
           form#options-form.form-horizontal
-        p #[b NOTE]: Installing may in rare cases lead to data loss. If the device is running Android, please note that you won't be able to access your Android data and/or apps in Ubuntu Touch. However the Android data will use storage space, so if you don't want it, please select the "Wipe" option. It's always wise to backup before installing.
```
Comment (NeoTheThird) [inline src/html/modals/options.pug]: Undo this.

File: src/html/modals/options.pug [hunk 11]
```diff
@@ -50,7 +49,7 @@
       _subdiv.appendChild(_input);
       _div.appendChild(_subdiv);
       $("#options-form").append(_div);
-
+
```
Comment (NeoTheThird) [inline src/html/modals/options.pug]: Undo this.

File: src/html/modals/options.pug [hunk 12]
```diff
@@ -77,19 +76,17 @@
         } else {
           ipcRenderer.send("option", option.var, $("#options-" + option.var).val());
         }
+        modals.show('install');
```
Comment (NeoTheThird) [inline src/html/modals/options.pug]: Undo this.

File: src/html/modals/options.pug [hunk 13]
```diff
@@ -77,19 +76,17 @@
...
       });
     }
 
     let optionsAdded = false;
     ipcRenderer.on("user:configure", (event, osInstructs) => {
+      views.show("working", "particles");
+      footer.underText.set("Please Configure the installation");
+      modals.show("options");
```
Comment (NeoTheThird) [inline src/html/modals/options.pug]: Undo this.

File: src/html/modals/options.pug [hunk 14]
```diff
@@ -77,19 +76,17 @@
...
       if (!optionsAdded) {
         optionsAdded = true;
         osInstructs.forEach(addOption);
       }
-      views.show("working", "particles");
-      footer.underText.set("Please Configure the installation");
-      modals.show("options");
-      $("#btn-options-close").click(() => {
-        setTimeout(() => ipcRenderer.send('install'), 250);
-      });
```
Comment (NeoTheThird) [inline src/html/modals/options.pug]: Undo this.

File: src/html/modals/udev.pug [hunk 15]
```diff
@@ -7,13 +7,11 @@
         h4.modal-title Warning!
       .modal-body
         p
-          | The installer needs to have the appropriate udev rules set to be able to communicate with devices over adb or fastboot. Click the button below to have the installer set up the rules automatically. You will be prompted to authorize this with your admin password.
-          br
-          | Alternatively, you can choose to do this #[a(onclick="shell.openExternal('https://docs.ubports.com/en/latest/userguide/install.html#missing-udev-rules')") manually] through the command-line.
+          | Please make sure you have the #[a(onclick="shell.openExternal('https://docs.ubports.com/en/latest/userguide/install.html#missing-udev-rules')") required udev-rules] set, or the Installer will not be able to detect your device.
+          | If you have already set the required rules, you can dismiss this message and simply continue with the installation.
```
Comment (NeoTheThird) [inline src/html/modals/udev.pug]: Undo this.

File: src/html/modals/udev.pug [hunk 16]
```diff
@@ -7,13 +7,11 @@
...
       .modal-footer
         button.btn.btn-default(type='button', data-dismiss='modal') Dismiss
-        button.btn.btn-default(type='button', data-dismiss='modal', onclick="localStorage.setItem('neverAskForUdevRules', 'true');") Don't ask me again
-        button.btn.btn-primary(type='button', data-dismiss='modal', onclick="localStorage.setItem('neverAskForUdevRules', 'true'); ipcRenderer.send('udev');") Set up rules automatically.
+        button.btn.btn-primary(type='button', data-dismiss='modal', onclick="localStorage.setItem('neverAskForUdevRules', 'true');") Don't ask me again
```
Comment (NeoTheThird) [inline src/html/modals/udev.pug]: Undo this.

File: src/html/views/select-os.pug [hunk 17]
```diff
@@ -15,8 +15,11 @@
           .col-xs-9
             select#options-os.form-control.space
       button#btn-inst.btn.btn-primary(type='button', style='width: 100%; margin-top: 10px;', onclick="ipcRenderer.send(\"os:selected\", getElementById('options-os').selectedIndex); global.installConfig.os_to_install = getElementById('options-os').selectedIndex;") Install
+      button#btn-backup-device.btn.btn-default(type='button', style='width: 100%; margin-top: 50px;', onclick="Backup();") Backup my device
+      button#btn-restore-device.btn.btn-default(type='button', style='width: 100%; margin-top: 10px;', onclick="Restore();") Restore a backup
   script.
     ipcRenderer.on("user:os", (event, installConfig, osSelects) => {
+      //global.utils.log.info("Test");
```
Comment (NeoTheThird) [inline src/html/views/select-os.pug]: Undo this.

File: src/main.js [hunk 18]
```diff
@@ -91,7 +91,7 @@ cli
   .option("-c, --cli", "[experimental] Run without GUI", undefined, "false")
   .option("-v, --verbose", "Enable verbose logging", undefined, "false")
   .option("-V, --veryVerbose", "Log *everything*", undefined, "false")
-  .option("-D, --debug", "Enable debugging tools", undefined, "false")
+  .option("-D, --debug", "Enable debugging tools", undefined, "true")
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/main.js [hunk 19]
```diff
@@ -200,23 +241,13 @@ ipcMain.on("os:selected", (event, osIndex) => {
     "user:configure",
     global.installConfig.operating_systems[osIndex]
   );
-  if (global.installConfig.operating_systems[osIndex].prerequisites.length) {
-    mainWindow.webContents.send(
-      "user:prerequisites",
-      global.installConfig,
-      osIndex
-    );
-  }
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/main.js [hunk 20]
```diff
@@ -200,23 +241,13 @@ ipcMain.on("os:selected", (event, osIndex) => {
...
 });
 
 // The user selected an os
 ipcMain.on("option", (event, targetVar, value) => {
   global.installProperties.settings[targetVar] = value;
 });
 
-// The user requested udev rules to be set
-ipcMain.on("udev", utils.setUdevRules);
-
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/main.js [hunk 21]
```diff
@@ -254,7 +285,6 @@ mainEvent.on("user:error", (error, restart, ignore) => {
 
 // Connection to the device was lost
 mainEvent.on("user:connection-lost", reconnect => {
-  utils.log.warn("connectiion to device lost");
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/main.js [hunk 22]
```diff
@@ -374,9 +521,6 @@ mainEvent.on("device", device => {
       global.installConfig,
       devices.getOsSelects(global.installConfig.operating_systems)
     );
-    if (global.installConfig.unlock.length) {
-      mainWindow.webContents.send("user:unlock", global.installConfig);
-    }
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/main.js [hunk 23]
```diff
@@ -568,6 +712,33 @@ app.on("ready", function() {
...
+            electron.shell.openExternal(
+              "https://docs.ubports.com/en/latest/userguide/install.html#troubleshooting"
+            )
+        },
+        {
+          label: "UBports Forums",
+          click: () => electron.shell.openExternal("https://forums.ubports.com")
+        }
+      ]
+    },
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/main.js [hunk 24]
```diff
@@ -598,55 +769,6 @@ app.on("ready", function() {
...
-        {
-          label: "Report a bug",
-          click: () => utils.sendBugReport("user-requested bug-report")
-        },
-        {
-          label: "Clean cached files",
-          click: utils.cleanInstallerCache
-        }
-      ]
-    },
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/main.js [hunk 25]
```diff
@@ -598,55 +769,6 @@ app.on("ready", function() {
...
-          click: () =>
-            electron.shell.openExternal(
-              "https://docs.ubports.com/en/latest/userguide/install.html#troubleshooting"
-            )
-        },
-        {
-          label: "UBports Forums",
-          click: () => electron.shell.openExternal("https://forums.ubports.com")
-        }
-      ]
```
Comment (NeoTheThird) [inline src/main.js]: Undo this.

File: src/system-image.js [hunk 26]
```diff
@@ -27,7 +27,10 @@ const getDeviceChannels = device => {
 };
 
 var installLatestVersion = options => {
-  return Promise.all([
```
Comment (NeoTheThird) [inline src/system-image.js]: undo everything in this file :)

File: src/utils.js [hunk 27]
```diff
@@ -19,7 +19,6 @@ const http = require("request");
 const download = require("download");
 const shell = require("electron").shell;
 const exec = require("child_process").exec;
-const sudo = require("sudo-prompt");
```
Comment (NeoTheThird) [inline src/utils.js]: Undo this.

File: src/utils.js [hunk 28]
```diff
@@ -196,25 +200,6 @@ function getLatestInstallerVersion() {
...
-      name: "UBports Installer",
-      icns: path.join(__dirname, "../build/icons/icon.icns")
-    },
-    error => {
-      if (error) log.warn("setting udev rules failed");
-      else log.debug("udev rules set");
-    }
-  );
-}
-
```
Comment (NeoTheThird) [inline src/utils.js]: Undo this.

File: src/utils.js [hunk 29]
```diff
@@ -244,8 +229,67 @@ function getUbuntuTouchDir() {
   return path.join(osCacheDir, "ubports");
 }
 
-function cleanInstallerCache() {
-  fs.emptyDir(getUbuntuTouchDir());
```
Comment (NeoTheThird) [inline src/utils.js]: Keep this function

File: src/utils.js [hunk 30]
```diff
@@ -261,11 +305,13 @@ let toolpath = global.packageInfo.package
...
   : path.join(__dirname, "..", "platform-tools", platforms[os.platform()]);
 let processes = [];
+
+// AW : Modifying the maxBuffer size to handle the output of the backup (all file path backuped on the device!)
 function execTool(tool, args, callback) {
   let pid = exec(
     [path.join(toolpath, tool)].concat(args).join(" "),
     {
-      maxBuffer: 1024 * 1024 * 2
+      maxBuffer: 1024 * 1024 * 100
```
Comment (NeoTheThird) [inline src/utils.js]: Not a good idea to do it like this. It'd be better to tar on the device and adb pull.

File: src/utils.js [hunk 31]
```diff
@@ -302,7 +353,25 @@ function killSubprocesses() {
 }
 
 function isSnap() {
-  return process.env.SNAP_NAME || false;
+  return process.env.SNAP_NAME;
```
Comment (NeoTheThird) [inline src/utils.js]: Undo this.

File: src/utils.js [hunk 32]
```diff
@@ -326,18 +395,14 @@ function checksumFile(file) {
                 "checked: " + path.basename(file.url),
                 sum === file.checksum
               );
-              if (sum === file.checksum) {
-                resolve();
-              } else {
+              if (sum === file.checksum) resolve();
+              else
```
Comment (NeoTheThird) [inline src/utils.js]: Undo this.

File: src/utils.js [hunk 33]
```diff
@@ -326,18 +395,14 @@ function checksumFile(file) {
...
-                      " but expected " +
-                      file.checksum
-                  )
+                  "checksum mismatch: calculated " +
+                    sum +
+                    " but expected " +
+                    file.checksum
                 );
-              }
             }
```
Comment (NeoTheThird) [inline src/utils.js]: Undo this.

File: src/utils.js (hunk 30 above)
Comment (Alain94W) [reply src/utils.js]: For now I leave it like this, I will take a look on how to handle this in a better way.

File: src/devices.js [hunk 34]
```diff
@@ -143,7 +143,7 @@ function installStep(step) {
     case "fastboot:erase":
       return new Promise(function(resolve, reject) {
         global.mainEvent.emit("user:write:working", "particles");
-        global.mainEvent.emit("user:write:status", "Cleaning up", true);
```
[the remark this answers is not in the document]
Comment (Alain94W) [reply src/devices.js]: Ok

File: src/devices.js [hunk 35]
```diff
@@ -176,10 +176,6 @@ function installStep(step) {
       });
     case "systemimage":
       return new Promise(function(resolve, reject) {
-        mainEvent.emit("user:write:progress", 0);
-        mainEvent.emit("user:write:working", "particles");
-        mainEvent.emit("user:write:status", "Downloading Ubuntu Touch", true);
-        mainEvent.emit("user:write:under", "Checking local files");
```
[the remark this answers is not in the document]
Comment (Alain94W) [reply src/devices.js]: ok

File: src/utils.js (hunk 30 above)
Comment (NeoTheThird) [reply src/utils.js]: Makes sense, especially since you need it to continue testing your current state over at your promise-android-tools pr :)

**Comment by NeoTheThird** [comment]
We have some merge conflicts now, you might want to consider taking some of your changes that can be easily grouped into separate commits (eg your pull animation) and applying them to the current state of master and getting them in separately. Else, these conflicts could get messy soon.

**Comment by Alain94W** [comment]
> We have some merge conflicts now, you might want to consider taking some of your changes that can be easily grouped into separate commits (eg your pull animation) and applying them to the current state of master and getting them in separately. Else, these conflicts could get messy soon.

Oh, and what about if I push all my work now and only disable the backup/restore option until I think about the best way of doing it in promise-android ? I mean I push all my work and on the select-os.pug I disable the Backup and restore ? Will there still an issue with the merging ?

**Comment by NeoTheThird** [comment]
@Alain94W 

> Oh, and what about if I push all my work now and only disable the backup/restore option until I think about the best way of doing it in promise-android ?

Please do! It's always good if i'm able to look at your code :)

> I mean I push all my work and on the select-os.pug I disable the Backup and restore ? Will there still an issue with the merging ?

we'll still have to resolve the conflicts, but i can help you with that. i appreciate that you're a git newbie, and rebasing, merging and cherry-picking can be intimidating at times :) I'm here to help

**Comment by NeoTheThird** [comment]
Closed in favor of #1088

**Pull request closed without merging**